docs/design/docker-deploy.md

```diff
@@ -0,0 +1,651 @@
+1. Deploying to a Docker Machine can be local or remote.
+1. Deploys can be made to different environments (dev, stage, prod) on the same or different Docker Machines.
+1. Image pushes can be made to Docker Hub, AWS ECR, and other public/private registries.
+1. Adhere to [Docker security best practices](https://docs.docker.com/engine/security/).
+1. Use sensible defaults to make it easy for users to get started.
+1. Utilize ngConfig project config file to save information for successive `docker:*` command runs.
+1. Support `--dry-run` and `--verbose` flags.
+
+
+## Proposed CLI API
+
+### Overview
+
+Initialize Docker files and configuration for a default build and run environment:
+```
+$ ng docker:init
+```
+
+Initialize a docker-compose file for deploying to a separate environment:
+```
+$ ng docker:init {environment}
+```
+
+Build and push a Docker image of the Angular app to the registry:
+```
+$ ng docker:push
+```
+
+Deploy and run the Angular app on a Docker Machine:
+```
+$ ng docker:deploy [environment]
+```
+
+Display logs of a running deployment:
+```
+$ ng docker:logs [environment] [--services service]
+```
+
+Attach to a running service container and run a command:
+```
+$ ng docker:exec [environment] [--service service] [--cmd bash]
+```
+
+### Command - Initialize Default Build and Run Environment
+
+The command `docker:init` prompts the user for configuration options and generates starter `Dockerfile`, `.dockerignore`, and `docker-compose.yml` files for building/pushing images to a registry (via `docker:push`) and running (via `docker:deploy`).
+
+Most users will start with one local 'default' Docker Machine (Mac/Win), or a local native Docker environment on Linux, where they will perform builds and run containers for development testing. This command prepares the Angular project for working with that default environment.
+
+Example - Init default build environment:
+```bash
+$ ng docker:init
+> Which Docker Machine for the default "build and run" environment?
+  [*] default / local
+  [ ] stage
+  [ ] production
+> What is the name of the Angular service container? [my-ng-app]
+> Which host port will the Angular service use in the default "build and run" environment? [8000]
+> What is the Docker registry address? [registry.hub.docker.com]
+> What is the registry image name (ie. "username/ng-app")? username/my-ng-app
+
+Generated 'Dockerfile'
+Generated '.dockerignore'
+Generated 'docker-compose.yml'
+
+Docker is ready!
+
+You can build and push a Docker image of your application to the registry using:
+  $ ng docker:push
+
+Build and run your application using:
+  $ ng docker:deploy
+```
+
+#### Options
+
+Arg / Prompt | Default | Variable name
+--- | --- | ---
+Which Docker Machine for the "build and run" environment?| *(Select from `docker-machine ls` names)* | defaultMachineName
+What is the name of the Angular service container? | *(`package.json` app name?)* | defaultServiceName
```

Comment:
When the user create a project he has to pass in a name that we will keep in the configuration. You could use that (it's going to be no space, kebab case).